Fix build limit icon bug

dnqbob opened the pull request:
Fix build limit icon bug:

1. Fix build limit performance bug:
    1.1 the icon will turn grey only when build limit is 1,  which means it won't work for build limit > 1
    1.2 the icon will change to avaliable palette when unit enter the transport

2. Add a "Isdead" check for build limit check in ProductionQueue.cs. I guess those fixs can prevent some bugs that not yet found.

Airdrop-product build limit related fix

1. when unit is being delivered, a placeholder will take up build limit to fix icon and buildable bug.

2. when airdrop-pad is destroyed and a transport is in delivering, the fix here will restore the current building number after the transport find it is not able to land and deliver.

3. when airdrop-pad is destroyed and a transport is in delivering, the fix here will refund the unit just built after the transport find it is not able to land and deliver.

4. ProductionQueue.cs build limit check is now using the function in TechTree.cs to improve performance and readability.

Fixes https://github.com/OpenRA/OpenRA/issues/5872.

dnqbob committed:
Fix build limit performance bug

1. Fix build limit performance bug when build limit:
    1.1 the icon will turn grey only when build limit is 1
    1.2 the icon will change to avaliable palette when unit enter the transport

2. Add a "Isdead" check for build limit check in ProductionQueue.cs

dnqbob commented:
There are some opinions about producing line & cargo:

1. I don't think directly remove passengers from world is a good idea, especially for AI squad behavior which is now in developing. Transport combats will be an important part for AI squad.

2. I believe it will be better for airdrop producing line (or any other similar producing line) give a place holder for unit that has already build and on its way to spawn point (can be a pre-load cargo or sth). If the building time is short enough for this unit which has build limit, player can build more than this limit. Any bug can be abused will be abused by player to gain advantage.

update: 2. is fixed

Mailaender commented on OpenRA.Mods.Common/Traits/Player/TechTree.cs:
```diff
@@ -94,14 +94,13 @@ public bool HasPrerequisites(IEnumerable<string> prerequisites)
 			}
 
 			// Add buildables that have a build limit set and are not already in the list
```
This comment is wrong now as you removed
```csharp
!ret.ContainsKey(a.Actor.Info.Name)
```

dnqbob replied:
Thanks for point out, I will correct it.

dnqbob commented:
Airdrop-product build limit related fix

1. when unit is being delivered,  the unit in airdrop transport will take up build limit.

2. when airdrop-pad is destroyed and a transport is in delivering , the logic here will restore the current building number after the transport find it is not able to land and deliver.

dnqbob committed:
Airdrop-produce related fix

1. when unit is being delivered, the unit in airdrop transport will take up build limit.
2. when airdrop-pad is destroyed and a transport is in delivering, the fix here will restore the current building number after the transport find it is not able to land and deliver.
3. when airdrop-pad is destroyed and a transport is in delivering, the fix here will refund after the transport find it is not able to land and deliver.
4. ProductionQueue.cs build limit check is now using the function in
TechTree to improve performance and readability.

dnqbob commented:
There are still a bug found by me about build limit icon bug and unfixed in this pr:

1. Pre-load passengers with build limit & provide prerequisites.
When actor is produced as pre-load passengers, its prerequisites and build limit won't work (don't trigger the `update()` in techtree) until the first time it get out of transport (once it get out of transport for the first time,  its prerequisites and build limit will work normally)

This one is pretty tricky which means we have to look into cargo while there are even two cargo type as carryall cargo and normal cargo (I also think these two should combined as one), in addition, pre-load passengers can also contain pre-load units which means we have to check recursively when unit is being producing and produced meanwhile many other edgecases are still unknown.

dnqbob commented:
Sorry for changing here, but use a place holder actor can cause many edgecase bug when they has a global effect (for example, it has support power, it can provide/consume power or it can provide prerequisites). 

Now fixed in my pr here.

dnqbob commented:
> when airdrop-pad is destroyed and a transport is in delivering, the fix here will refund the unit just built after the transport find it is not able to land and deliver.

This is a probably a balance fix in my thought. Airdrop is cool but in many ways disadvantage to normal production line. The worst one I think is when Airdrop platform is gone, the incoming plane will not either return cash or land on other Airdrop platform, instead they just fly away. I guess a refund is the most simple one for me to do.

pchote commented:
My concern here is that the changes look super risky (which is reinforced by having undergone several major revisions), but with limited payoff because it is still based on the legacy tech tree implementation.

We have a long-standing plan to rewrite the prequisites system on top of `VariableExpression` to allow proper relationships between different variables (e.g. limit the number of aircraft to the number of airfields), and I had always expected that #5872 would be addressed as part of that work.

dnqbob commented:
Could you please tell me where is the super risky change here? Do you mean breaking the encapsulation?

pchote commented:
The `TechTree` code is very fragile and confusingly mixes the concepts of build-limited units with prerequisites. There is a history of it breaking in unexpected ways whenever somebody tries to touch it. Having production icons not grey out in specific situations is a relatively benign bug compared to some of the previous issues.

dnqbob commented:
Ummm...  So would you care a test here to figure out if there are crashes will still happen here?  I already have some test at modified D2K and a localization mod I work on,  and it turns out good (for now). 

update: I am look into #5913 to confirm some of the bugs.

dnqbob commented:
So there is some problem may(will)  happen if I turn on "IsInWorld". 

The actors which provides prerequisites is dead/sold will not trigger the tech tree update,  which will be a mess to gameplay. 

I will try if I can fix it.

dnqbob commented:
@pchote 
Sorry to bother again but I have a question here,  why you do not remove the build limits check in tech tree since it is always buggy,  and just let `ProductionQueue` to do the job to check build limits?

I think turn grey for icon is only work for build limits = 1 and must not in transport are enough to make this turn grey function for build limits useless at most of situations,  remove it can be better, meanwhile save a system cost of looking for actors in world.

dnqbob commented:
pr closed.  I guess we need a discussion on issue.
